File: sdks/python/blocks_network/file_upload.py
"""
Shared pre-signed URL upload helper.

Implements the three-step upload handshake:
1. POST /api/v1/files/request-upload -> get uploadSessionId, uploadId, uploadUrl, formFields
2. POST file data to uploadUrl (direct S3 upload via multipart/form-data)
3. POST /api/v1/files/confirm-upload -> confirmation

Used by both TaskClient (consumer input) and AgentInstance (provider output).
"""

from __future__ import annotations

import io
import json
import ssl
import urllib.error
import urllib.request
from typing import Any, Dict, List, Optional, Tuple

import certifi

from .protocol_version import CURRENT_PROTOCOL_VERSION, PROTOCOL_VERSION_HEADER
from .write_affinity import capture_affinity, inject_affinity


class FileUploadError(Exception):
    """Raised when a file upload step fails."""

    def __init__(self, message: str, status_code: Optional[int] = None) -> None:
        super().__init__(message)
        self.status_code = status_code


def _build_multipart_body(
    form_fields: List[Dict[str, str]],
    file_data: bytes,
    file_name: str,
    mime_type: str,
) -> Tuple[bytes, str]:
    """Build a multipart/form-data body with form fields followed by the file.

    Returns ``(body_bytes, content_type_header)`` where the content type
    includes the boundary string.
    """
    boundary = "----BlocksUploadBoundary"
    parts: list[bytes] = []

    for field in form_fields:
        key = field.get("key", field.get("name", ""))
        value = field.get("value", "")
        parts.append(
            f"--{boundary}\r\n"
            f'Content-Disposition: form-data; name="{key}"\r\n'
            f"\r\n"
            f"{value}\r\n".encode("utf-8")
        )

    # File part must be last (S3 pre-signed POST requirement)
    parts.append(
        f"--{boundary}\r\n"
        f'Content-Disposition: form-data; name="file"; filename="{file_name}"\r\n'
        f"Content-Type: {mime_type}\r\n"
        f"\r\n".encode("utf-8")
    )
    parts.append(file_data)
    parts.append(f"\r\n--{boundary}--\r\n".encode("utf-8"))

    body = b"".join(parts)
    content_type = f"multipart/form-data; boundary={boundary}"
    return body, content_type


def _authenticated_json_post(
    url: str,
    body: Dict[str, Any],
    agent_auth: Any = None,
    auth_provider: Any = None,
) -> Dict[str, Any]:
    """POST JSON to a backend endpoint with authentication.

    Returns the parsed JSON response body.
    """
    # Pre-flight: when the auth provider has a recorded permanent-refresh
    # error, attempt one reactive recovery. On failure the typed
    # AuthRefreshFailedError is raised so file uploads surface it instead
    # of an opaque 401 from the request-upload / confirm-upload endpoints.
    from .auth_provider import preflight_auth_or_raise
    preflight_auth_or_raise(auth_provider)

    payload = json.dumps(body).encode("utf-8")
    headers: Dict[str, str] = {
        "Content-Type": "application/json",
        PROTOCOL_VERSION_HEADER: CURRENT_PROTOCOL_VERSION,
    }

    if agent_auth is not None:
        resp_data, status = agent_auth.authenticated_request(
            url,
            method="POST",
            body=payload,
            headers=headers,
        )
        if isinstance(resp_data, dict):
            if resp_data.get("error"):
                raise FileUploadError(
                    resp_data.get("error", "Upload request failed"),
                    status_code=status,
                )
            return resp_data
        return {}

    def _build_auth_headers() -> Dict[str, str]:
        hdrs = dict(headers)
        if auth_provider is not None:
            auth_header = auth_provider.get_auth_header()
            if auth_header:
                hdrs["Authorization"] = auth_header
        inject_affinity(hdrs)
        return hdrs

    def _execute(hdrs: Dict[str, str]) -> Dict[str, Any]:
        ssl_ctx = ssl.create_default_context(cafile=certifi.where())
        req = urllib.request.Request(url, data=payload, headers=hdrs, method="POST")

        try:
            with urllib.request.urlopen(req, context=ssl_ctx, timeout=30) as resp:
                capture_affinity(resp)
                resp_body = resp.read().decode("utf-8")
                return json.loads(resp_body) if resp_body else {}
        except urllib.error.HTTPError as exc:
            error_body = ""
            try:
                error_body = exc.read().decode("utf-8")
            except Exception:
                pass
            raise FileUploadError(
                f"HTTP {exc.code}: {error_body or exc.reason}",
                status_code=exc.code,
            ) from exc
        except urllib.error.URLError as exc:
            raise FileUploadError(str(exc.reason)) from exc

    req_headers = _build_auth_headers()
    try:
        return _execute(req_headers)
    except FileUploadError as err:
        # 401 reactive refresh: retry once if auth_provider can refresh
        if err.status_code == 401 and auth_provider is not None:
            if auth_provider.on_auth_failure():
                return _execute(_build_auth_headers())
        raise


def request_upload(
    base_url: str,
    *,
    role: str,
    file_name: str,
    file_size: int,
    mime_type: str,
    agent_name: Optional[str] = None,
    task_id: Optional[str] = None,
    part_id: Optional[str] = None,
    output_id: Optional[str] = None,
    upload_session_id: Optional[str] = None,
    agent_auth: Any = None,
    auth_provider: Any = None,
) -> Dict[str, Any]:
    """Step 1: Request a pre-signed upload URL from the backend.

    Parameters
    ----------
    base_url:
        Backend base URL.
    role:
        ``"consumer-input"`` or ``"provider-output"``.
    file_name, file_size, mime_type:
        File metadata.
    agent_name:
        Required for consumer-input (first file in session).
    task_id:
        Required for provider-output.
    part_id:
        Maps to ``requestPart.partId`` (consumer input).
    output_id:
        Maps to ``io.outputs[].id`` (provider output).
    upload_session_id:
        Pass to join an existing upload session (consumer input, additional files).
    agent_auth:
        AgentAuth instance for automatic token refresh.

    Returns
    -------
    dict
        Response with ``uploadSessionId``, ``uploadId``, ``uploadUrl``, ``formFields``.
        Provider output omits ``uploadSessionId``.
    """
    url = f"{base_url.rstrip('/')}/api/v1/files/request-upload"
    body: Dict[str, Any] = {
        "role": role,
        "fileName": file_name,
        "fileSize": file_size,
        "mimeType": mime_type,
    }
    if role == "consumer-input":
        if upload_session_id:
            body["uploadSessionId"] = upload_session_id
        elif agent_name:
            body["agentName"] = agent_name
        if part_id:
            body["partId"] = part_id
    elif role == "provider-output":
        if task_id:
            body["taskId"] = task_id
        if output_id:
            body["outputId"] = output_id

    return _authenticated_json_post(url, body, agent_auth=agent_auth, auth_provider=auth_provider)


def upload_to_presigned_url(
    upload_url: str,
    form_fields: List[Dict[str, str]],
    file_data: bytes,
    file_name: str,
    mime_type: str = "application/octet-stream",
) -> None:
    """Step 2: Upload file data directly to the pre-signed S3 URL.

    The upload is a standard S3 multipart POST with form fields first,
    then the file as the last part.

    Raises :class:`FileUploadError` on failure.
    """
    body, content_type = _build_multipart_body(form_fields, file_data, file_name, mime_type)

    ssl_ctx = ssl.create_default_context(cafile=certifi.where())
    req = urllib.request.Request(
        upload_url,
        data=body,
        headers={"Content-Type": content_type},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, context=ssl_ctx, timeout=120) as resp:
            # S3 returns 204 on success
            if resp.status not in (200, 204):
                raise FileUploadError(
                    f"Upload returned unexpected status {resp.status}",
                    status_code=resp.status,
                )
    except urllib.error.HTTPError as exc:
        error_body = ""
        try:
            error_body = exc.read().decode("utf-8")
        except Exception:
            pass
        raise FileUploadError(
            f"Upload failed: HTTP {exc.code}: {error_body or exc.reason}",
            status_code=exc.code,
        ) from exc
    except urllib.error.URLError as exc:
        raise FileUploadError(f"Upload failed: {exc.reason}") from exc


def confirm_upload(
    base_url: str,
    upload_id: str,
    *,
    agent_auth: Any = None,
    auth_provider: Any = None,
) -> Dict[str, Any]:
    """Step 3: Confirm the upload with the backend.

    For consumer input, returns ``{ uploadId }``.
    For provider output, returns ``{ uploadId, artifactRef }`` and the
    backend publishes the typed artifact event.
    """
    url = f"{base_url.rstrip('/')}/api/v1/files/confirm-upload"
    body = {"uploadId": upload_id}
    return _authenticated_json_post(url, body, agent_auth=agent_auth, auth_provider=auth_provider)


def presigned_upload_flow(
    base_url: str,
    file_data: bytes,
    *,
    role: str,
    file_name: str,
    mime_type: str = "application/octet-stream",
    agent_name: Optional[str] = None,
    task_id: Optional[str] = None,
    part_id: Optional[str] = None,
    output_id: Optional[str] = None,
    upload_session_id: Optional[str] = None,
    agent_auth: Any = None,
    auth_provider: Any = None,
) -> Dict[str, Any]:
    """Execute the full three-step pre-signed URL upload flow.

    Returns the combined result containing at least ``uploadId``.
    For consumer input, also contains ``uploadSessionId``.
    For provider output, also contains ``artifactRef``.
    """
    # Step 1: Request upload URL
    req_result = request_upload(
        base_url,
        role=role,
        file_name=file_name,
        file_size=len(file_data),
        mime_type=mime_type,
        agent_name=agent_name,
        task_id=task_id,
        part_id=part_id,
        output_id=output_id,
        upload_session_id=upload_session_id,
        agent_auth=agent_auth,
        auth_provider=auth_provider,
    )

    upload_url = req_result.get("uploadUrl", "")
    form_fields = req_result.get("formFields", [])
    upload_id = req_result.get("uploadId", "")

    if not upload_url or not upload_id:
        raise FileUploadError("request-upload did not return uploadUrl or uploadId")

    # Step 2: Direct upload to S3
    upload_to_presigned_url(upload_url, form_fields, file_data, file_name, mime_type)

    # Step 3: Confirm
    confirm_result = confirm_upload(
        base_url, upload_id,
        agent_auth=agent_auth,
        auth_provider=auth_provider,
    )

    # Merge request-upload and confirm-upload results
    merged: Dict[str, Any] = {}
    merged["uploadId"] = upload_id
    if "uploadSessionId" in req_result:
        merged["uploadSessionId"] = req_result["uploadSessionId"]
    if "artifactRef" in confirm_result:
        merged["artifactRef"] = confirm_result["artifactRef"]

    return merged

File: sdks/node/src/runtime/file-upload.ts
/**
 * Shared pre-signed URL upload helper for task artifacts.
 *
 * Implements the three-step upload handshake:
 *   1. request-upload  -> Backend returns uploadUrl, formFields, uploadId
 *   2. S3 multipart POST -> SDK uploads directly to PubNub Files storage
 *   3. confirm-upload  -> Backend validates and returns result
 *
 * Used by both consumer input (task-client.ts) and provider output
 * (agent-instance.ts). File bytes never flow through the backend.
 */

import { CURRENT_PROTOCOL_VERSION, PROTOCOL_VERSION_HEADER } from './protocol-version.js';
import type { AgentAuth } from './agent-auth.js';
import type { AuthProvider } from './auth-provider.js';
import { preflightAuthOrThrow } from './auth-provider.js';
import { captureAffinity, injectAffinity } from './write-affinity.js';

// ============================================================================
// Types
// ============================================================================

/** Fields returned by the backend for direct S3 upload. */
export interface UploadFormField {
  key: string;
  value: string;
}

/** Response from POST /api/v1/files/request-upload (consumer input). */
export interface RequestUploadResponseConsumer {
  uploadSessionId: string;
  uploadId: string;
  uploadUrl: string;
  formFields: UploadFormField[];
}

/** Response from POST /api/v1/files/request-upload (provider output). */
export interface RequestUploadResponseProvider {
  uploadId: string;
  uploadUrl: string;
  formFields: UploadFormField[];
}

/** Response from POST /api/v1/files/confirm-upload (consumer input). */
export interface ConfirmUploadResponseConsumer {
  uploadId: string;
}

/** Response from POST /api/v1/files/confirm-upload (provider output). */
export interface ConfirmUploadResponseProvider {
  uploadId: string;
  artifactRef: {
    kind: 'file';
    channel: string;
    mimeType: string;
    size: number;
    fileId: string;
    fileName: string;
    hash?: string;
    expiresAt?: string;
  };
}

/** Parameters for a consumer-input request-upload call. */
export interface ConsumerUploadParams {
  role: 'consumer-input';
  agentName: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  partId: string;
  uploadSessionId?: string;
}

/** Parameters for a provider-output request-upload call. */
export interface ProviderUploadParams {
  role: 'provider-output';
  taskId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  outputId?: string;
}

/** Auth context for making backend HTTP calls. */
export interface FileUploadAuth {
  baseUrl: string;
  authProvider?: AuthProvider;
  agentAuth?: AgentAuth;
}

// ============================================================================
// Internal helpers
// ============================================================================

/**
 * Make an authenticated HTTP request to the backend.
 * Uses AgentAuth.authenticatedFetch when available, otherwise plain fetch
 * with AuthProvider header. Handles 401 reactive refresh for authProvider.
 */
async function backendFetch(
  auth: FileUploadAuth,
  path: string,
  body: Record<string, unknown>,
): Promise<Response> {
  // Pre-flight: when the auth provider has a recorded permanent-refresh
  // error, attempt one reactive recovery. On failure the typed
  // AuthRefreshFailedError is thrown so file uploads surface it instead
  // of an opaque 401 from the request-upload / confirm-upload endpoints.
  await preflightAuthOrThrow(auth.authProvider);

  const url = `${auth.baseUrl.replace(/\/+$/, '')}${path}`;

  // agentAuth injects/captures affinity inside authenticatedFetch; the
  // non-agentAuth path below handles it directly.
  //
  // `credentials: 'omit'` keeps the browser from auto-attaching the
  // dashboard session cookie when the SDK and backend share an origin
  // (e.g. https://app.blocks.ai). Without this, the backend's auth
  // middleware resolves via the cookie path before reaching the Bearer
  // path, populates `req.session`, and the CSRF middleware then rejects
  // the call as `CSRF_INVALID`. The Bearer JWT in `Authorization` is
  // the SDK's auth and is not CSRF-vulnerable. Node fetch has no cookie
  // jar, so this is a no-op in provider contexts.
  const buildInit = (includeAffinity: boolean): RequestInit => {
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
      [PROTOCOL_VERSION_HEADER]: CURRENT_PROTOCOL_VERSION,
    };
    const authHeader = auth.authProvider?.getAuthHeader();
    if (authHeader) {
      headers['Authorization'] = authHeader;
    }
    if (includeAffinity) injectAffinity(headers);
    return {
      method: 'POST',
      credentials: 'omit',
      headers,
      body: JSON.stringify(body),
    };
  };

  if (auth.agentAuth) {
    return auth.agentAuth.authenticatedFetch(url, buildInit(false));
  }

  const response = await fetch(url, buildInit(true));
  captureAffinity(response.headers);

  // 401 reactive refresh for authProvider
  if (response.status === 401 && auth.authProvider) {
    const refreshed = await auth.authProvider.onAuthFailure();
    if (refreshed) {
      const retryResponse = await fetch(url, buildInit(true));
      captureAffinity(retryResponse.headers);
      return retryResponse;
    }
  }

  return response;
}

// ============================================================================
// Step 1: Request Upload
// ============================================================================

/**
 * Request a pre-signed upload URL from the backend.
 * Returns the upload URL, form fields, upload ID, and (for consumer input)
 * the upload session ID.
 */
export async function requestUpload(
  auth: FileUploadAuth,
  params: ConsumerUploadParams,
): Promise<RequestUploadResponseConsumer>;
export async function requestUpload(
  auth: FileUploadAuth,
  params: ProviderUploadParams,
): Promise<RequestUploadResponseProvider>;
export async function requestUpload(
  auth: FileUploadAuth,
  params: ConsumerUploadParams | ProviderUploadParams,
): Promise<RequestUploadResponseConsumer | RequestUploadResponseProvider> {
  const response = await backendFetch(auth, '/api/v1/files/request-upload', params as unknown as Record<string, unknown>);

  if (!response.ok) {
    const text = await response.text().catch(() => '');
    throw new Error(`request-upload failed: HTTP ${response.status} ${text}`);
  }

  return response.json() as Promise<RequestUploadResponseConsumer | RequestUploadResponseProvider>;
}

// ============================================================================
// Step 2: Direct S3 Upload
// ============================================================================

/**
 * Upload file data directly to PubNub Files storage via S3 pre-signed URL.
 * Constructs a multipart/form-data body with all form fields first,
 * then appends the file as the last field named "file".
 *
 * Uses the native `FormData` + `Blob` types available in both Node 18+
 * and all modern browsers. `fetch` computes the multipart boundary
 * automatically -- do NOT set `Content-Type` manually.
 *
 * Returns void on success (S3 returns 204 No Content).
 * Throws on failure (e.g., 400 EntityTooLarge for oversized files).
 */
export async function uploadToStorage(
  uploadUrl: string,
  formFields: UploadFormField[],
  fileData: Uint8Array | ArrayBuffer | Blob,
  fileName: string,
  mimeType: string,
): Promise<void> {
  const formData = new FormData();

  // Pre-signed URL form fields first (order matters for some S3 configs).
  for (const field of formFields) {
    formData.append(field.key, field.value);
  }

  // File as the last field, named "file" (required by S3 pre-signed POST).
  // The `as never` narrows away TypeScript's stricter `BlobPart`
  // constraint (which excludes the runtime-permissive
  // `Uint8Array<ArrayBufferLike>`). At runtime the Blob constructor
  // accepts any ArrayBufferView or ArrayBuffer; TS 5.4's lib typings
  // are conservative.
  const blob =
    fileData instanceof Blob
      ? fileData
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      : new Blob([fileData as any], { type: mimeType });
  formData.append('file', blob, fileName);

  const response = await fetch(uploadUrl, {
    method: 'POST',
    body: formData,
    // Do NOT set Content-Type -- fetch sets the multipart boundary
    // automatically when body is FormData.
  });

  if (!response.ok) {
    const text = await response.text().catch(() => '');
    throw new Error(`S3 upload failed: HTTP ${response.status} ${text}`);
  }
}

// ============================================================================
// Step 3: Confirm Upload
// ============================================================================

/**
 * Confirm a completed upload with the backend.
 *
 * For consumer input: returns { uploadId }.
 * For provider output: the backend publishes the typed artifact event
 * on the task channel and returns { uploadId, artifactRef }.
 */
export async function confirmUpload(
  auth: FileUploadAuth,
  uploadId: string,
): Promise<ConfirmUploadResponseConsumer>;
export async function confirmUpload(
  auth: FileUploadAuth,
  uploadId: string,
  expectArtifactRef: true,
): Promise<ConfirmUploadResponseProvider>;
export async function confirmUpload(
  auth: FileUploadAuth,
  uploadId: string,
  expectArtifactRef?: boolean,
): Promise<ConfirmUploadResponseConsumer | ConfirmUploadResponseProvider> {
  void expectArtifactRef; // used only for type narrowing at callsites
  const response = await backendFetch(auth, '/api/v1/files/confirm-upload', { uploadId });

  if (!response.ok) {
    const text = await response.text().catch(() => '');
    throw new Error(`confirm-upload failed: HTTP ${response.status} ${text}`);
  }

  return response.json() as Promise<ConfirmUploadResponseConsumer | ConfirmUploadResponseProvider>;
}

// ============================================================================
// Full Upload Flow (convenience wrapper)
// ============================================================================

/**
 * Run the complete three-step upload flow for a single file.
 * Shared by consumer input and provider output paths.
 */
export async function uploadFile(
  auth: FileUploadAuth,
  params: ConsumerUploadParams,
  fileData: Uint8Array | ArrayBuffer | Blob,
): Promise<{ uploadSessionId: string; uploadId: string }>;
export async function uploadFile(
  auth: FileUploadAuth,
  params: ProviderUploadParams,
  fileData: Uint8Array | ArrayBuffer | Blob,
): Promise<ConfirmUploadResponseProvider>;
export async function uploadFile(
  auth: FileUploadAuth,
  params: ConsumerUploadParams | ProviderUploadParams,
  fileData: Uint8Array | ArrayBuffer | Blob,
): Promise<{ uploadSessionId?: string; uploadId: string; artifactRef?: ConfirmUploadResponseProvider['artifactRef'] }> {
  // Step 1: Request upload URL
  const uploadResponse = await requestUpload(auth, params as ConsumerUploadParams);

  // Step 2: Upload directly to storage
  await uploadToStorage(
    uploadResponse.uploadUrl,
    uploadResponse.formFields,
    fileData,
    params.fileName,
    params.mimeType,
  );

  // Step 3: Confirm upload
  if (params.role === 'provider-output') {
    const confirmResult = await confirmUpload(auth, uploadResponse.uploadId, true);
    return confirmResult;
  }

  const consumerResponse = uploadResponse as RequestUploadResponseConsumer;
  await confirmUpload(auth, uploadResponse.uploadId);
  return {
    uploadSessionId: consumerResponse.uploadSessionId,
    uploadId: uploadResponse.uploadId,
  };
}
